tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go

```diff
@@ -131,31 +144,134 @@ func (a *AwsProvider) CreateWindowsVM() (rerr error) {
+// after the instance is in running state. Exposing this function to be used in testing later.
+func (a *AwsProvider) GetPublicIP(instanceID string) (string, error) {
+	// Till the instance comes to running state we cannot get the public ip address associated with it.
+	// So, it's better to query the AWS api again to get the instance object and the ip address.
+	instance, err := a.GetInstance(instanceID)
+	if err != nil {
+		return "", err
+	}
+	return *instance.PublicIpAddress, nil
+}
+
+// GetPassword returns the password associated with the string. Exposing this to be used in tests later
+func (a *AwsProvider) GetPassword(instanceID string) (string, error) {
+	privateKey, err := ioutil.ReadFile(a.privateKeyPath)
+	if err != nil {
+		return "", err
+	}
+	privateKeyBytes := privateKey
+	decodedEncryptedData, err := a.getDecodedPassword(instanceID)
+	if err != nil {
+		return "", err
+	}
+	decryptedPwd, err := rsaDecrypt(decodedEncryptedData, privateKeyBytes)
+	if err != nil {
+		return "", err
+	}
+	return string(decryptedPwd), nil
+}
+
+// getDecodedPassword gets the decoded password from the AWS cloud provider API
+func (a *AwsProvider) getDecodedPassword(instanceID string) ([]byte, error) {
+	// The docs within the aws-sdk says
+	// `If you try to retrieve the password before it's available,
+	// 	the output returns an empty string. We recommend that you wait up to 15 minutes
+	// after launching an instance before trying to retrieve the generated password.`
+	// So let's have hard timeout of 15 minutes.
+	startTime := time.Now()
+	var decodedPassword []byte
+	for {
+		currTime := time.Since(startTime)
+		if currTime >= 15*time.Minute {
+			// Did not get password in 15 minute time limit, let's exit
+			return nil, errors.New("error getting password in 15 minutes")
+		}
+		pwdData, err := a.getPasswordData(instanceID)
+		if err != nil {
+			return nil, err
+		}
+		// This means password is not fully formed
+		if pwdData == nil || len(*pwdData.PasswordData) == 0 {
+			continue
+		}
+		// Decode password
+		decodedPassword, err = base64.StdEncoding.DecodeString(strings.Trim(*pwdData.PasswordData, "\r\n"))
+		if err != nil {
+			return nil, err
+		}
+		return decodedPassword, nil
+	}
+}
+
+// getPasswordData returns the password passworddataoutput, if this returns nil, the password is not yet generated for the
+// instance
+func (a *AwsProvider) getPasswordData(instanceID string) (*ec2.GetPasswordDataOutput, error) {
+	pwdData, err := a.EC2.GetPasswordData(&ec2.GetPasswordDataInput{
+		InstanceId: aws.String(instanceID),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error getting encrypted password from aws cloud provider with %v", err)
+	}
+	return pwdData, nil
+}
+
+// rsaDecrypt decyptes the encrypted password and returns it.
+// From AWS docs:
+// The keys that Amazon EC2 uses are 2048-bit SSH-2 RSA keys. You can have up to five thousand key pairs per Region.
```

Comment:
I suspect this is for keys created using their [API](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html#having-ec2-create-your-key-pair)